File: cpp/se/DomainKVCommandExtension.cpp
#include "se/DomainKVCommandExtension.hpp"
#include "se/Command.hpp"
#include "se/KVExtension.hpp"
#include "xml/XMLHelper.hpp"

namespace {
    KVExtension& kvExtension() {
        static KVExtension* kvExtension = new KVExtension();
        return *kvExtension;
    }
}; // anonymous namespace

DomainKVCommandExtension::DomainKVCommandExtension(
        const CommandType* commandType) :
        commandName(commandType->getCommandName()),
        kvLists()
{
}

void DomainKVCommandExtension::addToCommand(const Command &command) const
{
    XMLWriter* xmlWriter = command.getXmlWriter();
    DOMElement* extensionElement = command.getExtensionElement();
    DOMElement* commandElement = xmlWriter->appendChild(extensionElement,
            commandName, kvExtension().getURI());

    createKVListElements(xmlWriter, commandElement);
}

void DomainKVCommandExtension::createKVListElements(
      XMLWriter *xmlWriter,
      DOMElement *commandElement) const
{
   DOMElement *kvlistElement = NULL;
   ExtensionList::const_iterator kvListsIterator;

   for (kvListsIterator = kvLists.begin();
         kvListsIterator != kvLists.end();
         kvListsIterator++)
   {
      kvlistElement = xmlWriter->appendChild(commandElement, "kvlist",
            kvExtension().getURI());
      kvlistElement->setAttribute(XStr("name").str(), XStr(kvListsIterator->first).str());
      addItemsToList(xmlWriter, kvListsIterator->second, kvlistElement);
   }
}

void DomainKVCommandExtension::addItemsToList(
      XMLWriter *xmlWriter,
      const KeyValueList &keyValueList,
      DOMElement *kvlistElement) const
{
   DOMElement *element = NULL;
   KeyValueList::const_iterator kvItemIterator;

   for (kvItemIterator = keyValueList.begin();
         kvItemIterator != keyValueList.end();
         kvItemIterator++)
   {
      element = xmlWriter->appendChild(kvlistElement, "item");
      element->setAttribute(XStr("key").str(), XStr(kvItemIterator->first).str());
      element->setTextContent(XStr(kvItemIterator->second).str());
   }
}

File: src/main/java/com/ausregistry/jtoolkit2/se/generic/DomainKVCommandExtension.java
package com.ausregistry.jtoolkit2.se.generic;

import java.util.Map.Entry;
import java.util.TreeMap;

import org.w3c.dom.Element;

import com.ausregistry.jtoolkit2.se.Command;
import com.ausregistry.jtoolkit2.se.CommandExtension;
import com.ausregistry.jtoolkit2.se.ExtendedObjectType;
import com.ausregistry.jtoolkit2.xml.XMLWriter;

/**
 * This class models the command elements as specified in the XML schema
 * definition urn:X-ar:params:xml:ns:kv-1.0.
 */
public final class DomainKVCommandExtension implements CommandExtension {

    private static final long serialVersionUID = -334068823995826478L;

    private final TreeMap<String, TreeMap<String, String>> kvLists;
    private final String commandType;

    /**
     * Creates a key-value domain extension for specified command type
     *
     * @param commandType
     *            the type of command
     *
     * @see com.ausregistry.jtoolkit2.se.CommandExtension#CREATE
     * @see com.ausregistry.jtoolkit2.se.CommandExtension#UPDATE
     */
    public DomainKVCommandExtension(final String commandType) {
        kvLists = new TreeMap<String, TreeMap<String, String>>();
        this.commandType = commandType;
    }

    public void addItem(final String listName, final String key, final String value) {
        TreeMap<String, String> kvList = kvLists.get(listName);

        if (kvList == null) {
            kvList = new TreeMap<String, String>();
            kvLists.put(listName, kvList);
        }

        kvList.put(key, value);
    }

    public void addToCommand(final Command command) {
        final XMLWriter xmlWriter = command.getXmlWriter();
        final Element extensionElement = command.getExtensionElement();
        final Element commandElement = createCommandElement(xmlWriter, extensionElement);
        createKVListElements(xmlWriter, commandElement);
    }

    private Element createCommandElement(final XMLWriter xmlWriter, final Element extensionElement) {
        final Element commandElement = xmlWriter.appendChild(extensionElement, commandType,
                ExtendedObjectType.KV.getURI());

        commandElement
                .setAttribute("xsi:schemaLocation", ExtendedObjectType.KV.getSchemaLocation());
        return commandElement;
    }

    private void createKVListElements(final XMLWriter xmlWriter, final Element commandElement) {
        for (final String kvListName : kvLists.keySet()) {
            final Element kvlistElement = xmlWriter.appendChild(commandElement, "kvlist",
                    ExtendedObjectType.KV.getURI());
            kvlistElement.setAttribute("name", kvListName);

            addItemsToList(xmlWriter, kvListName, kvlistElement);
        }
    }

    private void addItemsToList(final XMLWriter xmlWriter, final String kvListName,
            final Element kvlistElement) {
        Element element;

        for (final Entry<String, String> item : kvLists.get(kvListName).entrySet()) {
            element = xmlWriter.appendChild(kvlistElement, "item");
            element.setAttribute("key", item.getKey());
            element.setTextContent(item.getValue());
        }
    }
}
